### PR TITLE
fix: add upgrade logic to inject a kubectl deployer when an old kustomize deployer is detected

### PR DESCRIPTION
**Pull request by renzodavid9**
Fixes: #8423

**Description**
New logic to create a kubectl deployer when an old kustomize deployer is present during upgrade. The logic creates a new kubectl deployer and merge the values from the old kustomize deployer into it. When a kubectl deployer is already present, we try to merge it with the configuration from the old kustomize deployer. Code will error if the `defaultNamespace` or `flags.disableValidation` are present in both and are different.

### CODE REVIEW

File: pkg/skaffold/schema/v2beta29/upgrade.go [hunk 1]
```diff
@@ -118,6 +118,17 @@ func (opu *OnePipelineUpgrader) upgradeOnePipeline(oldPipeline, newPipeline inte
...
 		}
+
+		// Inject Kubectl deployer to deploy kustomize manifests
+		kubectlDeployerWasInjected := false
+		if newPL.Deploy.KubectlDeploy == nil {
+			newPL.Deploy.KubectlDeploy = &next.KubectlDeploy{}
+			kubectlDeployerWasInjected = true
+		}
+
+		if err := mergeKustomizeIntoKubectlDeployer(newPL.Deploy.KubectlDeploy, oldPL.Deploy.KustomizeDeploy, kubectlDeployerWasInjected); err != nil {
```
Comment (gsquared94) [inline pkg/skaffold/schema/v2beta29/upgrade.go:129]: will doing `mergeKustomizeIntoKubectlDeployer(oldPL *Pipeline, newPL *next.Pipeline)` simplify this?

File: pkg/skaffold/schema/v2beta29/upgrade.go [hunk 2]
```diff
@@ -300,3 +311,32 @@ func upgradePatches(olds []JSONPatch, news []next.JSONPatch) {
 		}
 	}
 }
+
+func mergeKustomizeIntoKubectlDeployer(newKubectlDeployer *next.KubectlDeploy, oldKustomizeDeployer *KustomizeDeploy, kubectlDeployerWasInjected bool) error {
```
Comment (gsquared94) [inline pkg/skaffold/schema/v2beta29/upgrade.go:315]: a good reading about variable names [here](go/go-style/decisions#variable-names)

Comment (renzodavid9) [reply pkg/skaffold/schema/v2beta29/upgrade.go]: Awesome! Thanks for the reference, I changed the variables name. Let me know what do you think.

File: pkg/skaffold/schema/v2beta29/upgrade.go (hunk 1 above)
Comment (renzodavid9) [reply pkg/skaffold/schema/v2beta29/upgrade.go]: Done! I moved all the logic to the new function I created, removing the third parameter used before. Let me know what do you think.

**Review by gsquared94 — APPROVED** [review]

**Comment by BabuRajan002** [comment]
We upgraded our Skaffold verstion to latest version 2.3 and Max Schema version is v4beat4. 
After that it stopped working during the deployment. Here is the sample file we are using
```
apiVersion: skaffold/v4beta4
kind: Config
metadata:
  name: sample-app
manifests:
  kustomize:
    paths:
      - </path/base>
profiles:
  - name: test
    manifests:
      kustomize:
        paths:
          - <path/kustomize>
        defaultNamespace: <dummy-namespace>
``` 

Our Kustomization's max schema version is "kustomize.config.k8s.io/v1beta1". Please help us to fix this issue. The current version we are using is 1.39 which accepts the above skaffold file with the kustomization version.

**Comment by renzodavid9** [comment]
Hey @BabuRajan002, could you please open a new issue in the [Skaffold repo](https://github.com/GoogleContainerTools/skaffold/issues) with the details of the error you are getting? Also, it will be great if you can include a repo with a minimal version that can be use to reproduce the problem. I think for this case the content of the Kustomization file will be good to understand what's happening. Also, we can use that new issue to continue the discussion. Thanks a lot!

**Comment by BabuRajan002** [comment]
@renzodavid9 As you mentioned above, I've created a new issue here - https://github.com/GoogleContainerTools/skaffold/issues/8802. But I did not get any response.

**Pull request merged**